File: src/main/java/org/jetbrains/teamcity/impl/fakes/FakeHttpSession.java


package org.jetbrains.teamcity.impl.fakes;

import org.jetbrains.annotations.NotNull;

import javax.servlet.ServletContext;
import javax.servlet.http.HttpSession;
import javax.servlet.http.HttpSessionContext;
import java.util.Enumeration;
import java.util.Hashtable;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

public class FakeHttpSession implements HttpSession {

  private static final AtomicInteger ourCounter = new AtomicInteger();

  private final String myId;
  private final long myCreationTime = System.currentTimeMillis();
  private final Map<String, Object> myAttrs = new ConcurrentHashMap<>();
  private volatile int myMaxInactiveInterval;
  private volatile ServletContext myServletContext;
  private volatile boolean myInvalidated;

  public FakeHttpSession() {
    myId = "" + ourCounter.incrementAndGet();
  }

  public void setServletContext(@NotNull ServletContext servletContext) {
    myServletContext = servletContext;
  }

  public long getCreationTime() {
    return myCreationTime;
  }

  public String getId() {
    return myId;
  }

  public long getLastAccessedTime() {
    return System.currentTimeMillis();
  }

  public ServletContext getServletContext() {
    return myServletContext;
  }

  public void setMaxInactiveInterval(int i) {
    myMaxInactiveInterval = i;
  }

  public int getMaxInactiveInterval() {
    return myMaxInactiveInterval;
  }

  public HttpSessionContext getSessionContext() {
    throw new UnsupportedOperationException("Not implemented in " + getClass().getName());
  }

  public Object getAttribute(String string) {
    checkValid();
    return myAttrs.get(string);
  }

  private void checkValid() {
    if (myInvalidated) {
      throw new IllegalStateException("Session invalidated");
    }
  }

  public boolean isInvalidated() {
    return myInvalidated;
  }

  public Object getValue(String string) {
    throw new UnsupportedOperationException("Not implemented in " + getClass().getName());
  }

  public Enumeration<String> getAttributeNames() {
    checkValid();
    return new Hashtable<>(myAttrs).keys();
  }

  public String[] getValueNames() {
    throw new UnsupportedOperationException("Not implemented in " + getClass().getName());
  }

  public void setAttribute(String string, Object object) {
    myAttrs.put(string, object);
  }

  public void putValue(String string, Object object) {
    throw new UnsupportedOperationException("Not implemented in " + getClass().getName());
  }

  public void removeAttribute(String string) {
    myAttrs.remove(string);
  }

  public void removeValue(String string) {
    throw new UnsupportedOperationException("Not implemented in " + getClass().getName());
  }

  public void invalidate() {
    myAttrs.clear();
    myInvalidated = true;
  }

  public boolean isNew() {
    throw new UnsupportedOperationException("Not implemented in " + getClass().getName());
  }
}

File: src/main/kotlin/org/jetbrains/teamcity/github/util/FakeHttpSession.kt


@file:Suppress("OverridingDeprecatedMember", "DEPRECATION")

package org.jetbrains.teamcity.github.util

import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger
import javax.servlet.ServletContext
import javax.servlet.http.HttpSession
import javax.servlet.http.HttpSessionContext

class FakeHttpSession : HttpSession {

    private val myId: String = "" + ourCounter.incrementAndGet()
    private val myCreationTime = System.currentTimeMillis()
    private val myAttrs = ConcurrentHashMap<String, Any>()
    @Volatile private var myMaxInactiveInterval: Int = 0
    @Volatile private var myServletContext: ServletContext? = null
    @Volatile
    var isInvalidated: Boolean = false
        private set

    override fun getCreationTime(): Long {
        return myCreationTime
    }

    override fun getId(): String {
        return myId
    }

    override fun getLastAccessedTime(): Long {
        return System.currentTimeMillis()
    }

    override fun getServletContext(): ServletContext? {
        return myServletContext
    }

    override fun setMaxInactiveInterval(i: Int) {
        myMaxInactiveInterval = i
    }

    override fun getMaxInactiveInterval(): Int {
        return myMaxInactiveInterval
    }

    override fun getSessionContext(): HttpSessionContext {
        throw UnsupportedOperationException("Not implemented in " + javaClass.name)
    }

    override fun getAttribute(string: String): Any? {
        checkValid()
        return myAttrs[string]
    }

    private fun checkValid() {
        if (isInvalidated) {
            throw IllegalStateException("Session invalidated")
        }
    }

    override fun getValue(string: String): Any {
        throw UnsupportedOperationException("Not implemented in " + javaClass.name)
    }

    override fun getAttributeNames(): Enumeration<String> {
        checkValid()
        return Hashtable(myAttrs).keys()
    }

    override fun getValueNames(): Array<String> {
        throw UnsupportedOperationException("Not implemented in " + javaClass.name)
    }

    override fun setAttribute(string: String, `object`: Any) {
        myAttrs[string] = `object`
    }

    override fun putValue(string: String, `object`: Any) {
        throw UnsupportedOperationException("Not implemented in " + javaClass.name)
    }

    override fun removeAttribute(string: String) {
        myAttrs.remove(string)
    }

    override fun removeValue(string: String) {
        throw UnsupportedOperationException("Not implemented in " + javaClass.name)
    }

    override fun invalidate() {
        myAttrs.clear()
        isInvalidated = true
    }

    override fun isNew(): Boolean {
        throw UnsupportedOperationException("Not implemented in " + javaClass.name)
    }

    companion object {
        private val ourCounter = AtomicInteger()
    }
}